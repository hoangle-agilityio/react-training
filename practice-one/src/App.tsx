import React from "react";
import TopNav from "./components/TopNav";

export default function App(): JSX.Element {
  return (
    <TopNav />
  );
}
