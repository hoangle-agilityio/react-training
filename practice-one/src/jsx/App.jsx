import React from "react";
import TopNav from "./components/TopNav";

export default function App() {
  return (
    <TopNav />
  );
}
