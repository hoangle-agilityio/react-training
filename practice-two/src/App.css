.app {
  background-color: #eeeeee;
  padding: 30px 0;
}

.app__heading {
  font-size: 36px;
  text-align: center;
  text-transform: uppercase;
  color: #333333;
  margin-bottom: 30px;
}

.user-list {
  border-collapse: collapse;
  width: 90%;
  margin: 0 auto;
}

.list-head,
.list-item {
  border: 1px solid #333333;
  text-align: left;
  padding: 8px;
}

.empty-item,
.error-msg {
  color: #ff0000;
  font-style: italic;
}

.create-user {
  width: 90%;
  margin: 0 auto;
  text-align: right;
}

.search-user {
  width: 90%;
  margin: 0 auto;
  text-align: right;
  margin-bottom: 30px;
}

.search-user input {
  padding: 5px;
  margin-left: 10px;
}

.loader {
  border: 16px solid #f3f3f3;
  border-radius: 50%;
  border-top: 16px solid #3498db;
  width: 120px;
  height: 120px;
  -webkit-animation: spin 2s linear infinite; /* Safari */
  animation: spin 2s linear infinite;
  position: absolute;
  top: 25%;
  left: 50%;
}

/* Safari */
@-webkit-keyframes spin {
  0% {
    -webkit-transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
  }
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
