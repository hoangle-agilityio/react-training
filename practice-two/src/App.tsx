import { useEffect, useState } from "react";
import "./App.css";
import Button from "./components/Button";
import ModalUser from "./components/Modal";
import { searching } from "./core/helpers/search-helper";
import User from "./core/interfaces/user";
import { deleteUser, getAllUsers } from "./services";

export default function App(): JSX.Element {
  const [users, setUsers] = useState<User[]>([]);
  const [currentUser, setCurrentUser] = useState<User>();
  const [isOpenModal, setIsOpenModal] = useState(false);
  const [searchTerm, setSearchTerm] = useState("");
  const [isViewUser, setIsViewUser] = useState(false);

  const handleOpenModal = (): void => setIsOpenModal(true);
  const handleCloseModal = (): void => setIsOpenModal(false);

  const openModalUser = (user?: User, isView: boolean = false) => {
    setIsViewUser(isView);
    setCurrentUser(user);
    handleOpenModal();
  }

  // Get all users
  const handleGetUsers = async (): Promise<void> => {
    try {
      setUsers(await getAllUsers());
    } catch (error) {
      throw new Error(`Get data failed: ${error}`);
    }
  };

  // Delete user when click to button delete
  const handleDeleteUser = async (user: User): Promise<void> => {
    if (confirm("Are you sure to delete this user?")) {
      try {
        await deleteUser(user.id);

        // Find index of user
        const index = users.indexOf(user);

        // If index exists, remove that index in array.
        // After that, set state user list after remove index
        if (index > -1) {
          users.splice(index, 1);
          setUsers([...users]);
        }

        alert("User deleted successfully!");
      } catch (error) {
        throw new Error(`Delete data failed: ${error}`);
      }
    }
  }

  // Filter user list by searchTerm
  const searchResult = users.filter(user => {
    return searching(user, searchTerm);
  });

  // Use useEffect avoid to call function repeatedly
  useEffect(() => {
    handleGetUsers();
  }, []);

  return (
    <div className="app">
      <h1 className="app__heading">User Management</h1>
      <section className="create-user">
        <Button
          buttonName="Add User"
          type="success"
          onClick={() => openModalUser(undefined)}
        />
      </section>
      <section className="search-user">
        <label htmlFor="search">Search:</label>
        <input type="text" id="search" onChange={event => setSearchTerm(event.target.value)} />
      </section>
      <section>
        <table className="user-list">
          <thead>
            <tr>
              <th className="list-head">User Id</th>
              <th className="list-head">Name</th>
              <th className="list-head">Email</th>
              <th className="list-head">Action</th>
            </tr>
          </thead>
          <tbody>
            {searchResult.length === 0 ? (
              <tr>
                <td className="empty-item">No data found!</td>
              </tr>
            ) : (
              <>
                {searchResult.map(user => (
                  <tr key={user.id}>
                    <td className="list-item">{user.id}</td>
                    <td className="list-item">{user.name}</td>
                    <td className="list-item">{user.email}</td>
                    <td className="list-item">
                      <Button
                        buttonName="View"
                        type="info"
                        onClick={() => openModalUser(user, true)}
                      />

                      <Button
                        buttonName="Edit"
                        type="primary"
                        onClick={() => openModalUser(user)}
                      />

                      <Button
                        buttonName="Delete"
                        type="danger"
                        onClick={() => handleDeleteUser(user)}
                      />
                    </td>
                  </tr>
                ))}
              </>
            )}
          </tbody>
        </table>
      </section>
      <ModalUser
        open={isOpenModal}
        currentUser={currentUser}
        users={users}
        isViewUser={isViewUser}
        onEdit={setIsViewUser}
        onSuccess={setUsers}
        closeModal={handleCloseModal}
      />
    </div>
  );
}
