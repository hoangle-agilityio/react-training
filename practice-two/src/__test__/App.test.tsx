import { cleanup, fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

// Create mock for app
jest.mock("../App");

// Create mock for modal
jest.mock("../components/Modal/index");

// Create mock for api
jest.mock("../services/index");

window.confirm = jest.fn(() => true);
window.alert = jest.fn();

describe('App Component', () => {
  afterEach(cleanup);

  it('should take a snapshot', () => {
    const { container } = render(<App />);
    expect(container.querySelector(".app")).toMatchSnapshot();
  });

  it('should be button Add User click event to open modal', () => {
    const { container } = render(<App />);
    const button = container.querySelector(".create-user button");
    const onClick = fireEvent.click(button!);

    expect(onClick).toBe(true);
  });

  it('should be button View User click event to open modal', () => {
    const { container } = render(<App />);
    const button = container.querySelector(".list-item .btn-info");
    const onClick = fireEvent.click(button!);

    expect(onClick).toBe(true);
  });

  it('should be button Edit User click event to open modal', () => {
    const { container } = render(<App />);
    const button = container.querySelector(".list-item .btn-primary");
    const onClick = fireEvent.click(button!);

    expect(onClick).toBe(true);
  });

  it('should be button click event to delete user', () => {
    const { container } = render(<App />);
    const button = container.querySelector(".btn-danger");
    const onClick = fireEvent.click(button!);

    expect(onClick).toBe(true);
  });

  it('should filter username with value entered', () => {
    const { container } = render(<App />);
    const search = container.querySelector("#search");
    
    userEvent.type(search!, "o");

    expect(container.querySelectorAll(".username").length).toBe(2);
  });
});
