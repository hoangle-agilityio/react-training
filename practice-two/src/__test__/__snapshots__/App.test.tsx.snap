// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should take a snapshot 1`] = `
<div
  class="app"
>
  <h1
    class="app__heading"
  >
    User Management
  </h1>
  <section
    class="create-user"
  >
    <button
      class="btn btn-success"
    >
      Add User
    </button>
  </section>
  <section
    class="search-user"
  >
    <label
      for="search"
    >
      Search:
    </label>
    <input
      id="search"
      type="text"
    />
  </section>
  <section>
    <table
      class="user-list"
    >
      <thead>
        <tr>
          <th
            class="list-head"
          >
            User Id
          </th>
          <th
            class="list-head"
          >
            Name
          </th>
          <th
            class="list-head"
          >
            Email
          </th>
          <th
            class="list-head"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="list-item"
          >
            1
          </td>
          <td
            class="list-item username"
          >
            Hoang
          </td>
          <td
            class="list-item"
          >
            hoang.le@asnet.com.vn
          </td>
          <td
            class="list-item"
          >
            <button
              class="btn btn-info"
            >
              View
            </button>
            <button
              class="btn btn-primary"
            >
              Edit
            </button>
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="list-item"
          >
            2
          </td>
          <td
            class="list-item username"
          >
            Roy Le
          </td>
          <td
            class="list-item"
          >
            royle@gmail.com
          </td>
          <td
            class="list-item"
          >
            <button
              class="btn btn-info"
            >
              View
            </button>
            <button
              class="btn btn-primary"
            >
              Edit
            </button>
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
  <section>
    <h2
      class="modal-title"
    >
      Add User
    </h2>
    <div
      class="input-group"
    >
      <label
        for="userName"
      >
        User Name
      </label>
      <input
        id="userName"
        placeholder="user name"
        type="text"
        value=""
      />
    </div>
    <div
      class="input-group"
    >
      <label
        for="userEmail"
      >
        User Email
      </label>
      <input
        id="userEmail"
        placeholder="user email"
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-success"
    >
      Add User
    </button>
    <button
      class="btn btn-secondary"
    >
      Cancel
    </button>
  </section>
</div>
`;
